### PR TITLE
[bser] add ability to encode to template

### PR DESCRIPTION
**Pull request by ShawnROGrady**
Added the ability to encode a slice/array of structs using the [more compact template format](https://facebook.github.io/watchman/docs/bser.html#array-of-templated-objects).  While this isn't 100% necessary, when compared to the standard encoding approach (`use_template=False`) using the template format is faster and uses less memory particularly as the length of the slice (`num_elems`) increases:

struct_arr_time-v-num_elems
struct_arr_mem_used-v-num_elems

### CODE REVIEW

File: bser/encode.go [hunk 1]
```diff
@@ -120,6 +120,9 @@ func encode(buf []byte, d interface{}) ([]byte, error) {
 			order.PutUint64(b, math.Float64bits(r.Float()))
 			return appendItem(buf, 0x07, b), nil
 		case reflect.Slice, reflect.Array:
+			if r.Type().Elem().Kind() == reflect.Struct {
```
Comment (jonasi) [inline bser/encode.go]: add a todo to handle pointers to structs

File: bser/encode.go [hunk 2]
```diff
@@ -176,6 +179,36 @@ func encode(buf []byte, d interface{}) ([]byte, error) {
...
+	elem := r.Type().Elem()
+	fieldNames := make([]string, elem.NumField())
+	b, err := encode(nil, r.Len())
+	if err != nil {
+		return nil, err
+	}
+	for i := 0; i < r.Len(); i++ {
+		s := r.Index(i)
+		for j := 0; j < s.NumField(); j++ {
+			if i == 0 {
```
Comment (jonasi) [inline bser/encode.go]: do we need to worry about unexpected fields?

File: bser/encode.go (hunk 1 above)
Comment (ShawnROGrady) [reply bser/encode.go]: @jonasi I'll need to double check but won't that already be handled? Looking at lines 103-111 it looks like for pointers we just call `encode()` on the dereferenced value which will be handled by this block if applicable.

```
elem := false
for r.Kind() == reflect.Ptr || r.Kind() == reflect.Interface {
	r = r.Elem()
	elem = true
}

if elem {
	return encode(buf, r.Interface())
}
```

File: bser/encode.go (hunk 2 above)
Comment (ShawnROGrady) [reply bser/encode.go]: @jonasi could you clarify what you mean by that? Are you referring to a situation where different structs in the array might have different fields (e.g. `[{A: 1, B: 2}, {C: 3, D:4}]`) or is there something else?

I don't think it is possible to have an array of different structs in go, unless maybe it is an array of interfaces, [in which case it wouldn't pass](https://play.golang.org/p/TfsJYHIxbio) the `if r.Type().Elem().Kind() == reflect.Struct`  check we have prior to calling `encodeTemplate(buf, r)`

Comment (jonasi) [reply bser/encode.go]: @ShawnROGrady I mean for a type like `struct { A string; b string }`. Do we have to worry about `b`?

Comment (ShawnROGrady) [reply bser/encode.go]: @jonasi good catch! Turns out we weren't handling this correctly in the non-template case either which resulted in a panic when a struct had an unexported field. Added a new `exportedStructFields()` helper to correctly handle this

Comment (ShawnROGrady) [reply bser/encode.go]: @jonasi moved those changes to a separate PR: https://github.com/jonasi/watchman/pull/21

Comment (jonasi) [reply bser/encode.go]: haha sorry - just realized I got autocorrected to `unexpected`.  obviously meant `unexported` in my first post...

**Pull request merged**